src/write_statsd.c

```diff
@@ -0,0 +1,476 @@
+  int idx;
+  int status;
+
+  for (idx = 0; idx < conf->children_num; idx++) {
+    oconfig_item_t *child = conf->children + idx;
+    status = 0;
+
+    if (strcasecmp("Host", child->key) == 0) {
+      status = cf_util_get_string(child, &configuration.host);
+
+    } else if (strcasecmp("Port", child->key) == 0) {
+      status = cf_util_get_int(child, & configuration.port);
+      
+    } else if (strcasecmp("Postfix", child->key) == 0) {
+      status = cf_util_get_string(child, &configuration.postfix);
+
+    } else if (strcasecmp("Prefix", child->key) == 0) {
+      status = cf_util_get_string(child, &configuration.prefix);
+    
+    } else if (strcasecmp("SilenceTypeWarnings", child->key) == 0) {
+      status = cf_util_get_boolean(
+          child, &configuration.silence_type_warnings);
+
+    } else if (strcasecmp("AlwaysAppendDS", child->key) == 0) {
+      status = cf_util_get_boolean(
+          child, &configuration.always_append_ds);
+
+    } else {
+      WARNING("write_statsd plugin: Ignoring unknown config option '%s'.",
+              child->key);
+    }
+
+    /*
+     * If any of the options cannot be parsed print an error message
+     * but keep going on with the defaults.
+     * If the host was not parsed correctily the plugin init step
+     * will fail for us.
+     */
+    if (status != 0) {
+      ERROR("write_statsd plugin: Ignoring config option '%s' due to an error.",
+            child->key);
+      return status;
+    }
+  }
+  return 0;
+}
+
+static int write_statsd_free(void) {
+  FREE_NOT_NULL(configuration.host);
+  FREE_NOT_NULL(configuration.postfix);
+  FREE_NOT_NULL(configuration.prefix);
+
+  configuration.event_line_base_len = 0;
+  configuration.short_event_line_base_len = 0;
+  FREE_NOT_NULL(configuration.event_line_format);
+  FREE_NOT_NULL(configuration.short_event_line_format);
+
+  DEBUG("Freed %s module.", WRITE_STATSD_NAME);
+  return 0;
+}
+
+static int write_statsd_init(void) {
+  size_t buffer_len = 0;
+  size_t len = 0;
+
+  if (configuration.host == NULL) {
+    ERROR("write_statsd plugin: missing required 'Host' configuration.");
+    FREE_NOT_NULL(configuration.postfix)
+    FREE_NOT_NULL(configuration.prefix)
+    return -1;
+  }
+
+  /*
+   * Generate format for event lines to pass to ssnprintf.
+   * A line full event line would be:
+   * <prefix>.<metric-name>.<value-name>.<postfix>:<value>|<type>
+   * The length of the format must exclude the placeholders.
+   *
+   * The value-name and its separator are ignored in the
+   * short format configuration.
+   */
+  if (configuration.prefix == NULL && configuration.postfix == NULL) {
+    /* %s.%s:%s|%s && %s:%s|%s */
+    configuration.event_line_format = strdup("%s.%s:%s|%s");
+    configuration.short_event_line_format = strdup("%s:%s|%s");
+    len = 3;
+
+  } else if (configuration.prefix == NULL) {
+    /* %s.%s.<postfix>:%s|%s && %s.<postfix>:%s|%s */
+    len = 4 + strlen(configuration.postfix);
```

Comment:
This is way too error prone. Please do the line formatting when writing metrics, ideally in a separate function, and don't store the format in a variable. The function could look something like this:

```c
int format_line (char *buffer, size_t buffer_size, int ds_type, value_t v)
{
  /* format v */
  if ((configuration.prefix != NULL) && (configuration.postfix != NULL))
    ssnprintf (buffer, buffer_size, "%s.%s.%s.%s:%s|%s",
        prefix, foo, bar, postfix, qux, v_str);
  else if (configuration.prefix != NULL)
    ssnprintf (buffer, buffer_size, "%s.%s.%s:%s|%s",
        prefix, foo, bar, qux, v_str);
  else
    /* … */
}
```

Yes, it's repetitive, but it's easy to read – maybe even boring – and readers don't have to wonder what the "base length" is supposed to be.